lib/lcevc/lcevc_dil.js

```diff
@@ -0,0 +1,277 @@
+ */
+
+goog.provide('shaka.lcevc.Dil');
+goog.require('shaka.log');
+goog.require('shaka.util.Dom');
+goog.require('shaka.util.IReleasable');
+
+/**
+ * @summary
+ * lcevcDil - (MPEG-5 Part 2 LCEVC - Decoder Integration Layer) provides
+ *  all the operations related to the enhancement and rendering
+ *  of LCEVC enabled streams and on to a canvas.
+ * @implements {shaka.util.IReleasable}
+ * @export
+ */
+shaka.lcevc.Dil = class {
+  /**
+   * @param {HTMLVideoElement} media The video element that will be attached to
+   * LCEVC Dil for input.
+   * @param {HTMLCanvasElement} canvas The canvas element that will be attached
+   * to LCEVC Dil to render the enhanced frames.
+   * @param {LcevcDil.LcevcDilConfig=} dilConfig The LCEVC DIL
+   * config object to initialize the LCEVC DIL.
+   */
+  constructor(media, canvas, dilConfig) {
+    /** @private {?LcevcDil.LcevcDIL} */
+    this.dil_ = null;
+
+    /** @private {?Element} */
+    this.parent_ = null;
+
+    /** @private {?HTMLElement} */
+    this.lcevcContainer_ = null;
+
+    /** @private {ResizeObserver} */
+    this.resizeObserver_ = null;
+
+    this.level_ = -1;
+    this.containerFormat_ = 0;
+
+    /** @type {HTMLVideoElement} */
+    this.media_ = media;
+
+    /** @type {HTMLCanvasElement} */
+    this.canvas_ = canvas;
+
+    /** @type {LcevcDil.LcevcDilConfig} */
+    this.dilConfig_ = dilConfig ? dilConfig :
+        shaka.lcevc.Dil.DefaultLcevcDilConfig;
+
+    this.createDil_();
+  }
+
+  /**
+   * Append data to the LCEVC Dil.
+   * @param {BufferSource} data
+   */
+  appendBuffer(data) {
+    if (this.dil_) {
+      this.dil_.appendBuffer(data, 'video', this.level_);
+    }
+  }
+
+  /**
+  * Hide the canvas and show original video element
+  */
+  hideCanvas() {
+    if (this.dil_) {
+      this.media_.style.display = 'block';
+      this.canvas_.style.display = 'none';
+    }
+  }
+
+  /**
+    * Show the canvas and hide original video element
+    */
+  showCanvas() {
+    if (this.dil_) {
+      this.media_.style.display = 'none';
+      this.canvas_.style.display = 'block';
+    }
+  }
+
+  /**
+   * Create LCEVC Dil.
+   * @private
+   */
+  createDil_() {
+    if (this.isSupported_() && !this.dil_) {
+      if (!this.canvas_) {
+        this.createCanvas_();
+      }
+      if (LcevcDil.SupportObject.webGLSupport(this.canvas_)) {
+        this.dil_ = new LcevcDil.LcevcDil(
+            this.media_,
+            this.canvas_,
+            this.dilConfig_);
+      }
+    }
+  }
+
+  /**
+   * Close LCEVC Dil.
+   * @override
+   * @export
+   */
+  release() {
+    this.hideCanvas();
+    if (this.dil_) {
+      this.dil_.close();
+      this.dil_ = null;
+    }
+    if (this.lcevcContainer_) {
+      this.lcevcContainer_.remove();
+    }
+    if (this.resizeObserver_) {
+      this.resizeObserver_.disconnect();
+      this.resizeObserver_ = null;
+    }
+    if (this.canvas_) {
+      this.canvas_.remove();
+    }
+  }
+
+  /**
+   * Check if LCEVC Dil exists.
```

Comment:
Nit: I would reword this JSDoc. It makes it sound like this checks if `this.dil_` is non-null.
Instead I'd say something like:
```js
/**
 * Check if the LVEVC code is present and supported by the browser.
 * @return {boolean}
 * @private
 */
```

Reply:
This message is changed from `Check if LCEVC Dil exists.` to `Check if the LCEVC Dil lib is present and is supported by the browser.`